yew/src/virtual_dom/vlist.rs

```diff
@@ -78,137 +173,320 @@ impl VDiff for VList {
+                .collect();
+
+            // We will try to match the left vnodes with the right ones, and
+            // store them in `matched_rights`: it is formed of vnodes from
+            // right, in the same order than left. `reused_rights` is used to
+            // know which right nodes have been matched and reused.
+            let mut matched_rights: Vec<Option<VNode>> = vec![None; self.children.len()];
+            let mut reused_rights = FixedBitSet::with_capacity(rights.len());
+
+            // Generator of dummy vnodes, used when stealing vnodes from rights.
+            let make_dummy = || {
+                VNode::VText(VText {
+                    text: String::default(),
+                    reference: None,
+                })
+            };
+
+            // Match all the elements that are not new, but may have moved.
+            let mut new_lefts: HashSet<usize> = HashSet::with_capacity(self.children.len());
+            for ((left_idx, left), matched_right) in self
+                .children
+                .iter()
+                .enumerate()
+                .zip(matched_rights.iter_mut())
+            {
+                let left_key = left.key().as_ref().expect("must have a key");
+                if let Some(right_idx) = rights_lookup.remove(&left_key) {
+                    let right = rights
+                        .get_mut(right_idx)
+                        .expect("the index from the map must be valid");
+                    let right = core::mem::replace(right, make_dummy());
+                    *matched_right = Some(right);
+                    reused_rights.put(right_idx);
                 } else {
-                    rights_lookup.insert(key.clone(), r);
+                    // This left node is new.
+                    new_lefts.insert(left_idx);
                 }
-            } else {
-                rights_nokeys.push(r);
             }
-        }
-        let mut rights = rights_nokeys.into_iter();
-        for left in lefts {
-            if let Some(key) = &left.key() {
-                match rights_lookup.remove(key) {
-                    Some(right) => {
-                        previous_sibling =
-                            left.apply(scope, parent, previous_sibling.as_ref(), Some(right));
+            let new_lefts = new_lefts; // remove mutability
+
+            // Reconciliation loop, i.e. apply the least amount of
+            // transformations to rights to make them identical to lefts.
+            let mut matched_rights = matched_rights.into_iter().peekable();
+            for left in self.children.iter_mut() {
+                let right = matched_rights.next().unwrap();
+
+                let next_sibling = matched_rights
+                    .peek()
+                    .and_then(|vnode| vnode.as_ref().and_then(|vnode| vnode.reference()));
+                let _ = left.apply(parent_scope, parent, next_sibling, right);
```

Comment:
If keyed right is not able to be matched with keyed left (ie. different types of node) we really need `next_sibling` to be reliable. I'm worried that the current implementation of `reference()` doesn't return a reference for `VList` and I'm not super confident that `VComp` will always have a reference when it should either

Reply:
> If keyed right is not able to be matched with keyed left (ie. different types of node)

I didn't think about this case. We should add tests then.